reconcile/utils/gitlab_api.py

```diff
@@ -686,6 +664,14 @@ def remove_labels(
         gitlab_request.labels(integration=INTEGRATION_NAME).inc()
         item.save()
 
+    @staticmethod
+    def is_group_project_owner(group_name: str, repo_url: str) -> bool:
+        """Utility function to check if the project is owned by the group by checking if group name part of repo url"""
+        substr = f"/{group_name}/"
+        if substr in repo_url:
+            return True
+        return False
```

Comment:
is there an official way to check group project owner instead of check url string?

Reply:
there should be a way but this is a hack....gitlab follows a standard url for repos i.e https://{instance}/{group_ OR_member}/{reponame}...It will save additional call and processing..

Reply:
see if any group info or project info can be used directly to check, and need to handle multi level urls, like `https://{instance}/{group}/{sub_group}/{repo}`

Reply:
@hemslo  the `get_all_group_projects`is a single  function will take care of this. It returns all the project inside group (owned+shared+inherited). In case of owned+inherited from parent we cannot share/unshare this project so we keep it as is... https://github.com/app-sre/qontract-reconcile/blob/f977b7143a732ad951456b4051189e4f74b626a5/reconcile/gitlab_permissions.py#L67

Reply:
I don't understand that logic, especially comments inside `extract_group_spec`, this worth doing a review session.

Reply:
Now we can simplify `extract_group_spec` using `next`, to indicate we always expect a match:

```python
    def extract_group_spec(self, project: GroupProject) -> GroupSpec:
        return next(
            GroupSpec(
                group_name=self.group.name,
                group_access_level=group["group_access_level"],
            )
            for group in project.shared_with_groups
            if group["group_id"] == self.group.id
        )
```

Then construct current state as:

```python
        current_state = {
            project.web_url: self.extract_group_spec(project)
            for project in shared_projects
        }
```

And since `group_name` is always `self.group.name`, it's not required to be in spec for this use case, but can be kept for later support multiple groups.